sdks/java/io/csv/src/main/java/org/apache/beam/sdk/io/csv/CsvIOParseHelpers.java

```diff
@@ -73,9 +75,34 @@ static void validate(CSVFormat format, Schema schema) {}
    * Build a {@link List} of {@link Schema.Field}s corresponding to the expected position of each
    * field within the CSV record.
    */
-  // TODO(https://github.com/apache/beam/issues/31718): implement method.
-  static List<Schema.Field> mapFieldPositions(CSVFormat format, Schema schema) {
-    return new ArrayList<>();
+  static Map<Integer, Schema.Field> mapFieldPositions(CSVFormat format, Schema schema) {
+    List<String> header = Arrays.asList(format.getHeader());
+    Map<Integer, Schema.Field> indexToFieldMap = new HashMap<>();
+    for (Schema.Field field : schema.getFields()) {
+      int index = getIndex(header, field);
+      if (index >= 0) {
+        indexToFieldMap.put(index, field);
+      }
+    }
+    return indexToFieldMap;
+  }
+
+  /**
+   * Attains expected index from {@link CSVFormat's} header matching a given {@link Schema.Field}.
+   */
+  private static int getIndex(List<String> header, Schema.Field field) {
+    String fieldName = field.getName();
+    boolean presentInHeader = header.contains(fieldName);
```

Comment:
list.contains is inefficient (O(n)). Instead of call getIndex in a for loop, how about changing 

```
List<String> header = Arrays.asList(format.getHeader());
```

to

```
HashSet<String> header = ...
```
where query from a hashset is O(1) ?

Reply:
HashSet is good for contains(), but is O(N) in order to find the indexOf(Schema.Field.getName()). This is done during pipeline construction, so performance will not be an issue.

Reply:
> This is done during pipeline construction, so performance will not be an issue.

I see, sounds good